controllers/toolchainconfig/configuration.go

```diff
@@ -1,15 +1,82 @@
+	"time"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	commonconfig "github.com/codeready-toolchain/toolchain-common/pkg/configuration"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	ToolchainStatusName = "toolchain-status"
+
+	// NotificationDeliveryServiceMailgun is the notification delivery service to use during production
+	NotificationDeliveryServiceMailgun = "mailgun"
+)
+
+var logger = logf.Log.WithName("toolchainconfig")
 
 type ToolchainConfig struct {
-	cfg *toolchainv1alpha1.ToolchainConfigSpec
+	cfg     *toolchainv1alpha1.ToolchainConfigSpec
+	secrets map[string]map[string]string
+}
+
+// GetToolchainConfig returns a ToolchainConfig using the cache, or if the cache was not initialized
+// then retrieves the latest config using the provided client and updates the cache
+func GetToolchainConfig(cl client.Client) (ToolchainConfig, error) {
+	config, secrets, err := commonconfig.GetConfig(cl, &toolchainv1alpha1.ToolchainConfig{})
+	if err != nil {
+		// return default config
+		logger.Error(err, "failed to retrieve ToolchainConfig")
+		return ToolchainConfig{cfg: &toolchainv1alpha1.ToolchainConfigSpec{}}, err
+	}
+	return newToolchainConfig(config, secrets), nil
+}
+
+// GetCachedToolchainConfig returns a ToolchainConfig directly from the cache
+func GetCachedToolchainConfig() ToolchainConfig {
+	config, secrets := commonconfig.GetCachedConfig()
+	return newToolchainConfig(config, secrets)
 }
 
-func NewToolchainConfig(cfg *toolchainv1alpha1.ToolchainConfigSpec) ToolchainConfig {
-	return ToolchainConfig{
-		cfg: cfg,
+// ForceLoadToolchainConfig updates the cache using the provided client and returns the latest ToolchainConfig
+func ForceLoadToolchainConfig(cl client.Client) (ToolchainConfig, error) {
+	config, secrets, err := commonconfig.LoadLatest(cl, &toolchainv1alpha1.ToolchainConfig{})
+	if err != nil {
+		// return default config
+		logger.Error(err, "failed to force load ToolchainConfig")
+		return ToolchainConfig{cfg: &toolchainv1alpha1.ToolchainConfigSpec{}}, err
 	}
+	return newToolchainConfig(config, secrets), nil
+}
+
+func newToolchainConfig(config runtime.Object, secrets map[string]map[string]string) ToolchainConfig {
+	if config == nil {
+		// return default config if there's no config resource
+		return ToolchainConfig{cfg: &toolchainv1alpha1.ToolchainConfigSpec{}}
+	}
+
+	toolchaincfg, ok := config.(*toolchainv1alpha1.ToolchainConfig)
+	if !ok {
+		// return default config
+		logger.Error(fmt.Errorf("cache does not contain toolchainconfig resource type"), "failed to get ToolchainConfig from resource, using default configuration")
+		return ToolchainConfig{cfg: &toolchainv1alpha1.ToolchainConfigSpec{}}
+	}
+	return ToolchainConfig{cfg: &toolchaincfg.Spec, secrets: secrets}
+}
+
+func (c *ToolchainConfig) Print() {
+	logger.Info("Toolchain configuration variables", "ToolchainConfigSpec", c.cfg)
```

Comment:
wouldn't make sense to convert it into a YAML format so it's easily parsable in case of later troubleshooting?

Reply:
How about if it's JSON? Like this:

```
2021-08-16T23:00:21.402Z	INFO	toolchainconfig	Toolchain configuration variables
{
	"host": {
		"automaticApproval": {
			"enabled": true,
			"resourceCapacityThreshold": {},
			"maxNumberOfUsers": {}
		},
		"deactivation": {
			"deactivationDomainsExcluded": "@redhat.com"
		},
		"metrics": {},
		"notifications": {
			"secret": {}
		},
		"registrationService": {
			"analytics": {},
			"auth": {},
			"environment": "dev",
			"verification": {
				"secret": {
					"ref": "host-operator-secret",
					"twilioAccountSID": "twilio.account.sid",
					"twilioAuthToken": "twilio.auth.token",
					"twilioFromNumber": "twilio.from_number"
				}
			}
		},
		"tiers": {},
		"toolchainStatus": {},
		"users": {}
	},
	"members": {
		"default": {
			"auth": {},
			"autoscaler": {
				"deploy": false
			},
			"che": {
				"secret": {}
			},
			"console": {},
			"memberStatus": {
				"refreshPeriod": "1s"
			},
			"toolchainCluster": {
				"healthCheckPeriod": "5s"
			},
			"webhook": {}
		}
	}
```

Reply:
JSON looks good :+1: